File: dotnet/source/cmf.bus/Envelope.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace cmf.bus
{
    /// <summary>
    /// Represents a discreet message sent across the bus. Envelopes consist of two parts, 
    /// payload represented as an array of bytes and optionally a set of headers represented 
    /// as string/string map of key/value pairs. Together they comprise the message and any 
    /// appropriate meta-data pertaining to it respectively.
    /// </summary>
    public sealed class Envelope
    {
        /// <summary>
        ///  The envelope's set of headers.
        /// </summary>
        public IDictionary<string, string> Headers { get; set; }

        /// <summary>
        /// The envelope's payload.
        /// </summary>
        public byte[] Payload { get; set; }

        /// <summary>
        /// Initializes a new instance of the Envelope class.
        /// </summary>
        public Envelope()
        {
            this.Headers = new Dictionary<string, string>();
        }

        /// <summary>
        /// Initializes a new instance of the Envelope class with specific headers and payload.
        /// </summary>
        /// <param name="headers">The headers for the envelope.</param>
        /// <param name="payload">The payload of the envelope.</param>
        public Envelope(IDictionary<string, string> headers, byte[] payload)
        {
            this.Headers = headers;
            this.Payload = payload;
        }

        /// <summary>
        /// Returns a string representation of the envelope that lists the key/value pairs contained in the envelope's headers.
        /// </summary>
        /// <returns>A string representation of the envelope.</returns>
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder("{");

            if (this.Headers.Count > 0)
            {
                KeyValuePair<string, string> first = this.Headers.First();
                sb.AppendFormat("\"{0}\":\"{1}\"", first.Key, first.Value);
 
                for (int index = 1; index < this.Headers.Count(); index++)
                {
                    sb.AppendFormat(",\"{0}\":\"{1}\"", first.Key, first.Value);
                }
            }

            sb.Append("}");

            return sb.ToString();
        }
    }
}

File: java/cmf-bus-api/src/main/java/cmf/bus/Envelope.java
package cmf.bus;


import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Represents a discreet message sent across the bus.  Envelopes consist of two 
 * parts, payload represented as an array of bytes and optionally a set of headers 
 * represented as string/string map of key/value pairs.  Together they comprise 
 * the message and any appropriate meta-data pertaining to it respectively.
 * @see IEnvelopeSender#send(Envelope)
 * @see IRegistration#handle(Envelope)
 */
public class Envelope {

    protected Map<String, String> headers = new HashMap<String, String>();

    protected byte[] payload = {};

    /**
     * @see java.lang.Object#equals(java.lang.Object)
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (!(obj instanceof Envelope)) {
            return false;
        }
        Envelope other = (Envelope) obj;
        if (!Arrays.equals(payload, other.payload)) {
            return false;
        } else if (!Arrays.equals(headers.values().toArray(), other.headers.values().toArray())) {
            return false;
        }

        return true;
    }

    /**
     * Gets the header value for the key specified.
     * @param key  The key that identifies the header value to return.
     * @return The value of the specified header.
     */
    public String getHeader(String key) {
        return headers.get(key);
    }

    /**
     * Gets all the envelope headers as a key-value map.  Caution: modifying the returned map
     * will modify the envelopes headers. 
     * @return A key/value map of headers.
     */
   public Map<String, String> getHeaders() {
        return headers;
    }

    /**
     * Gets the payload (i.e. the actual content) of the envelope.
     * @return The content of the envelope as a byte array.
     */
    public byte[] getPayload() {
        return payload;
    }

    /**
     * Sets a header value for the specified key.
     * @param key The key that identifies the header value to set.
     * @param value The value to set the header to.
     */
    public void setHeader(String key, String value) {
        headers.put(key, value);
    }

    /**
     * Sets all headers based on the provided map.  Calling this method will delete
     * any pre-existing headers.
     * 
     * @param headers A key/value map of headers for the envelop.
     */
    public void setHeaders(Map<String, String> headers) {
        this.headers = headers;
    }

    /**
     * Sets the payload of the envelope.
     * 
     * @param payload A byte array containing the payload of the envelope.
     */
    public void setPayload(byte[] payload) {
        if (payload == null) {
            throw new IllegalArgumentException("Envelope payload cannot be set to null");
        }
        this.payload = payload;
    }

    /**
     * Returns a string representation of the envelope that lists the key/value pairs 
     * contained in the envelope's headers.
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("{");

        boolean first = true;

        for (Map.Entry<String, String> header : this.headers.entrySet()) {
            if (!first) {
                sb.append(", ");
            }
            else {
                first = false;
            }

            sb.append("\"");
            sb.append(header.getKey());
            sb.append("\" : \"");
            sb.append(header.getValue());
            sb.append("\"");
        }

        sb.append("}");
        return sb.toString();
    }
}
